app/javascript/mastodon/actions/compose.js

```diff
@@ -390,13 +390,13 @@ export function hydrateCompose() {
   };
 }
 
-function insertIntoTagHistory(tags) {
+function insertIntoTagHistory(recognizedTags, text) {
   return (dispatch, getState) => {
     const state = getState();
     const oldHistory = state.getIn(['compose', 'tagHistory']);
     const me = state.getIn(['meta', 'me']);
-    const names = tags.map(({ name }) => name);
-    const intersectedOldHistory = oldHistory.filter(name => !names.includes(name));
+    const names = recognizedTags.map(tag => text.match(new RegExp(`#${tag.name}`, 'i'))[0].slice(1));
```

Comment:
Oh yeah, right, the hashtags shouldn't need to be escaped, valid hashtags are only word characters.
I can think of one last corner case, that is, the hashtag being a substring of another string in the status, for instance a URL hash… but I guess that's a rare corner case…